### DISCUSSION TITLE
useMutation - type errors with my mutation function

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [TanStack/query#1216] [Q&A]
Hey fellow react-query users,

I'm using Typescript with `react-query` and I'm getting type errors raised for the mutation function I'm passing into `useMutation`.

This is the typescript error I'm receiving:

```
Argument of type '(username: string, password: string) => Promise<CognitoUser>' is not assignable to parameter of type 'MutationFunction<CognitoUser, string>'
```

I created a super small [codesanbox](https://codesandbox.io/s/silly-satoshi-q8p9q?file=/src/App.tsx:0-990) of my code. Two of the `useMutation` examples are throwing type errors but the 3rd one isn't (_using a much simpler function_)

Any ideas on how to work around this?  Any help is appreciated 😄 
![Screen Shot 2020-10-25 at 10 53 32 PM](https://user-images.githubusercontent.com/6743796/97138230-f14be600-1714-11eb-902d-00809e3d7f66.png)

[Code Sanbox Link](https://codesandbox.io/s/silly-satoshi-q8p9q?file=/src/App.tsx:0-990)

```tsx
//imports....

class AuthenticationService {
  static async signin(username: string, password: string){
      const AWSCognitoUser = await Auth.signIn({username,password}) as CognitoUser;
      return AWSCognitoUser;
  }
  static async deleteSomething(username:string){
    return "SUCCESS"
  }
}

export default function App() {
  // Example1: type errors
  const [signin1] = useMutation(AuthenticationService.signin)
  
  // Example2  - type errors
  const [signin2] = useMutation(async (username:string, password:string):Promise<CognitoUser> => {
      return await AuthenticationService.signin(username, password)
  })

  // Example3, no type errors
  const [deleteUser] = useMutation(AuthenticationService.deleteSomething)

  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
    </div>
  );
}

```

**Comment by courier-new** [comment]
Hi @cliffordfajardo! It's quite possible you solved this by now, but for anyone else who might happen across this discussion thread, I wanted to share the realization I had when I ran into this!

A `MutationFunction` only accepts **one argument**, which is why in your example `deleteSomething(username)` works fine but `signin(username, password)` does not. You can see this from the [full type definition](https://github.com/tannerlinsley/react-query/blob/master/src/core/types.ts#L266) too:

```typescript
export type MutationFunction<TResult, TVariables = unknown> = (
  variables: TVariables
) => Promise<TResult>
```

It's easy to miss that when the type gets collapsed to `MutationFunction<CognitoUser, string>` in the error message, though. 😄 

The way around this is pass the variables to the mutation as object, as it [recommends in the docs](https://react-query.tanstack.com/docs/guides/mutations#mutation-variables) (silly me forgot about this little important detail). So:

```typescript
const signin = ({ username, password}: {username: string, password: string}) => {
    const AWSCognitoUser = await Auth.signIn({username,password}) as CognitoUser;
    return AWSCognitoUser;
}

const [signin1] = useMutation(signin)

// Invoke later
const me = signin({username: "super-user", password: "password1234"})
```

Hope this might help! 🙂

**Comment by xegulon** [comment]
This is not emphasized enough in the docs @courier-new , that you have to give it only one argument

**Reply by TkDodo** [reply]
feel free to PR to the docs. It's mentioned on the [mutations guide](https://tanstack.com/query/v4/docs/guides/mutations):

> In the example above, you also saw that you can pass variables to your mutations function by calling the mutate function with a **single variable or object**.

But I can see that the example uses a single string variable, and maybe it isn't clear that it _cannot_ be multiple variables.
